src/promql/src/planner.rs

```diff
@@ -0,0 +1,627 @@
+                })
+            }
+            PromExpr::MatrixSelector { .. } => UnsupportedExprSnafu {
+                name: "Prom Matrix Selector",
+            }
+            .fail()?,
+            PromExpr::Call { func, args } => {
+                let args = self.create_function_args(args)?;
+                let input =
+                    self.prom_expr_to_plan(args.input.with_context(|| ExpectExprSnafu {
+                        expr: prom_expr.clone(),
+                    })?)?;
+                let mut func_exprs = self.create_function_expr(func, args.literals)?;
+                func_exprs.insert(0, self.create_time_index_column_expr()?);
+                LogicalPlanBuilder::from(input)
+                    .project(func_exprs)
+                    .context(DataFusionSnafu)?
+                    .build()
+                    .context(DataFusionSnafu)?
+            }
+        };
+        Ok(res)
+    }
+
+    fn selector_to_series_normalize_plan(
+        &self,
+        table_name: &str,
+        offset: Option<Duration>,
+        label_matchers: Matchers,
+    ) -> Result<LogicalPlan> {
+        // TODO(ruihang): add time range filter
+        let filter = self.matchers_to_expr(label_matchers)?;
+        let table_scan = self.create_relation(table_name, filter)?;
+        let offset = offset.unwrap_or_default();
+
+        let series_normalize = SeriesNormalize::new(
+            offset,
+            self.ctx
+                .time_index_column
+                .clone()
+                .context(NoTimeIndexSnafu)?,
+            table_scan,
+        );
+        let logical_plan = LogicalPlan::Extension(Extension {
+            node: Arc::new(series_normalize),
+        });
+        Ok(logical_plan)
+    }
+
+    // TODO(ruihang): ignore `MetricNameLabel` (`__name__`) matcher
+    fn matchers_to_expr(&self, label_matchers: Matchers) -> Result<Vec<DfExpr>> {
+        let mut exprs = Vec::with_capacity(label_matchers.matchers.len());
+        for matcher in label_matchers.matchers {
+            let col = DfExpr::Column(Column::new(None::<String>, matcher.name));
+            let lit = DfExpr::Literal(ScalarValue::Utf8(Some(matcher.value)));
+            let expr = match matcher.op {
+                MatchOp::Equal => col.eq(lit),
+                MatchOp::NotEqual => col.not_eq(lit),
+                MatchOp::Re(_) => DfExpr::BinaryExpr(BinaryExpr {
+                    left: Box::new(col),
+                    op: Operator::RegexMatch,
+                    right: Box::new(lit),
+                }),
+                MatchOp::NotRe(_) => DfExpr::BinaryExpr(BinaryExpr {
+                    left: Box::new(col),
+                    op: Operator::RegexNotMatch,
+                    right: Box::new(lit),
+                }),
+            };
+            exprs.push(expr);
+        }
+
+        Ok(exprs)
+    }
+
+    fn create_relation(&self, table_name: &str, filter: Vec<DfExpr>) -> Result<LogicalPlan> {
```

Comment:
What about `create_table_scan_plan`? It looks better.

Reply:
I copy this name from data fusion's planner. But when afterward there are many differences that prevented me from keeping copying. So this name becomes the wired one. I'll change it tomorrow